haxe/ui/core/Component.hx

```diff
@@ -1746,74 +1732,285 @@ class Component extends ComponentBase implements IComponentBase implements IClon
+
+        for (flag in _invalidationFlags.keys()) {
+            _invalidationFlags.remove(flag);
+        }
+
+        _isAllInvalid = false;
+
+        for (flag in _delayedInvalidationFlags.keys()) {
+            if (flag == InvalidationFlags.ALL) {
+                _isAllInvalid = true;
+            } else {
+                _invalidationFlags.set(flag, true);
+            }
+            _delayedInvalidationFlags.remove(flag);
+        }
+        _isValidating = false;
+    }
+
+    private function validateInternal():Void {
+        var styleInvalid = isInvalid(InvalidationFlags.STYLE);
+        var positionInvalid = isInvalid(InvalidationFlags.POSITION);
+        var displayInvalid = isInvalid(InvalidationFlags.DISPLAY);
+        var layoutInvalid = isInvalid(InvalidationFlags.LAYOUT) && _layoutLocked == false;
+
+        if (styleInvalid) {
+            validateStyle();
+        }
+
+        if (hasTextDisplay()) {
+            getTextDisplay().validate();
+        }
+
+        if (hasTextInput()) {
+            getTextInput().validate();
+        }
+
+        //TODO
+        //if (hasImageDisplay()) {
+        //    getImageDisplay().validate();
+        //}
+
+        if (positionInvalid) {
+            validatePosition();
+
+            onMoved();
+            dispatch(new UIEvent(UIEvent.MOVE));
+        }
+
+        if (layoutInvalid) {
+            validateLayout();
         }
 
-        _layoutInvalidating = true;
+        if (displayInvalid || styleInvalid || layoutInvalid) {
+            var oldComponentWidth = _componentWidth;
+            var oldComponentHeight = _componentHeight;
+
+            if (layoutInvalid || styleInvalid) {
+                validateAutoSize();
+            }
+
+            validateDisplay();
+
+            if(displayInvalid || oldComponentWidth != _componentWidth || oldComponentHeight != _componentHeight) {
+                onResized();
+                dispatch(new UIEvent(UIEvent.RESIZE));
+            }
+        }
+    }
 
+    private function validateLayout():Void {
         layout.refresh();
+    }
 
-        _layoutInvalidating = false;
+    private function validateStyle():Void {
+        var s:Style = Toolkit.styleSheet.applyClasses(this, false);
+        if (_ready == false || _style == null || _style.equalTo(s) == false) { // lets not update if nothing has changed
+            _style = s;
+            applyStyle(_style);
+        }
+    }
 
-        if (_layoutReinvalidation == true) {
-            _layoutReinvalidation = false;
-            invalidateLayout();
+    private function validatePosition():Void {
+        handlePosition(_left, _top, _style);
+    }
+
+    private function validateDisplay():Void {
+        if (componentWidth == null || componentHeight == null || componentWidth <= 0 || componentHeight <= 0) {
+            return;
+        }
+
+        handleSize(componentWidth, componentHeight, _style);
+    }
+
+    private function validateAutoSize():Void {
```

Comment:
I see, so this is where autosize ends up? So it might just be a matter of this isnt called at the right points?